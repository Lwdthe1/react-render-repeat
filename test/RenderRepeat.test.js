import React from "react";
import pjson from "../package.json";
import path from "path";
import renderer from "react-test-renderer";

// Test the root of the lib
const RenderRepeat = require(path.join(__dirname, "../", pjson.main)).default;

test("Example 1", () => {
  const list = [
    { id: "uA1", name: "User A" },
    { id: "uB2", name: "User B" },
    { id: "uC3", name: "User C" }
  ];

  const component = renderer.create(
    <RenderRepeat
      list={list}
      render={({ el, index, isLast }) => {
        return (
          <div key={el.id}>
            <p>ID: {el.id}</p>
            <p>Name: {el.name}</p>
            {isLast ? <p>As index {index}, I'm the last in the list</p> : null}
          </div>
        );
      }}
    />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("Example 2: Render with placeholders", () => {
  const list = [
    { id: "uA1", name: "User A" },
    { id: "uB2", name: "User B" },
    { id: "uC3", name: "User C" },
    { $isPlaceholder: true },
    { $isPlaceholder: true },
    { $isPlaceholder: true },
    { $isPlaceholder: true },
    { $isPlaceholder: true }
  ];

  const component = renderer.create(
    <RenderRepeat
      list={list}
      render={({ el, index, isLast }) => {
        return (
          <div key={el.id}>
            <p>ID: {el.id}</p>
            <p>Name: {el.name}</p>
            {isLast ? <p>As index {index}, I'm the last in the list</p> : null}
          </div>
        );
      }}
      isPlaceholder={({ $isPlaceholder }) => !!$isPlaceholder}
      renderPlaceholder={({ index }) => (
        <div className="u-relative" key={`placeholder${index}`}>
          <p>I am the #{index} loading indicator ... </p>
          <p>
            I should be replaced after the real data returns from the server.
          </p>
        </div>
      )}
    />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("Example 3: Render with placeholders with isLast check", () => {
  const list = [
    { id: "uA1", name: "User A" },
    { id: "uB2", name: "User B" },
    { $isPlaceholder: true },
    { $isPlaceholder: true },
    { $isPlaceholder: true }
  ];

  const renderListElement = ({ el, index }) => {
    return <p key={el.id}>Element at index {index}</p>;
  };

  const checkElementIsPlaceholder = ({ $isPlaceholder }) => !!$isPlaceholder;

  const renderPlaceholderElement = ({ index, isLast }) => (
    <p key={`placeholder${index}`}>
      Placeholder at index {index}{" "}
      {isLast ? <span>is last!</span> : <span>is not last.</span>}
    </p>
  );

  const component = renderer.create(
    <RenderRepeat
      list={list}
      render={renderListElement}
      isPlaceholder={checkElementIsPlaceholder}
      renderPlaceholder={renderPlaceholderElement}
    />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
