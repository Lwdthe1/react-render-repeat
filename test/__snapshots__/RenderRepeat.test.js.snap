// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example 1 1`] = `
Array [
  <div>
    <p>
      ID: 
      uA1
    </p>
    <p>
      Name: 
      User A
    </p>
  </div>,
  <div>
    <p>
      ID: 
      uB2
    </p>
    <p>
      Name: 
      User B
    </p>
  </div>,
  <div>
    <p>
      ID: 
      uC3
    </p>
    <p>
      Name: 
      User C
    </p>
    <p>
      As index 
      2
      , I'm the last in the list
    </p>
  </div>,
]
`;

exports[`Example 2: Render with placeholders 1`] = `
Array [
  <div>
    <p>
      ID: 
      uA1
    </p>
    <p>
      Name: 
      User A
    </p>
  </div>,
  <div>
    <p>
      ID: 
      uB2
    </p>
    <p>
      Name: 
      User B
    </p>
  </div>,
  <div>
    <p>
      ID: 
      uC3
    </p>
    <p>
      Name: 
      User C
    </p>
  </div>,
  <div
    className="u-relative"
  >
    <p>
      I am the #
      3
       loading indicator ... 
    </p>
    <p>
      I should be replaced after the real data returns from the server.
    </p>
  </div>,
  <div
    className="u-relative"
  >
    <p>
      I am the #
      4
       loading indicator ... 
    </p>
    <p>
      I should be replaced after the real data returns from the server.
    </p>
  </div>,
  <div
    className="u-relative"
  >
    <p>
      I am the #
      5
       loading indicator ... 
    </p>
    <p>
      I should be replaced after the real data returns from the server.
    </p>
  </div>,
  <div
    className="u-relative"
  >
    <p>
      I am the #
      6
       loading indicator ... 
    </p>
    <p>
      I should be replaced after the real data returns from the server.
    </p>
  </div>,
  <div
    className="u-relative"
  >
    <p>
      I am the #
      7
       loading indicator ... 
    </p>
    <p>
      I should be replaced after the real data returns from the server.
    </p>
  </div>,
]
`;

exports[`Example 3: Render with placeholders with isLast check 1`] = `
Array [
  <p>
    Element at index 
    0
  </p>,
  <p>
    Element at index 
    1
  </p>,
  <p>
    Placeholder at index 
    2
     
    <span>
      is not last.
    </span>
  </p>,
  <p>
    Placeholder at index 
    3
     
    <span>
      is not last.
    </span>
  </p>,
  <p>
    Placeholder at index 
    4
     
    <span>
      is last!
    </span>
  </p>,
]
`;
